File: 03-classic-algorithms/tree/binarytree/binary_tree.js
/**
 * Copyright © https://github.com/jarry All rights reserved.
 * @author: jarryli@gmail.com
 * @version: 1.0
 */

// 1. 先序优先遍历DLR递归版
function preOrderTraverse(tree, result) {
  result = result ? result : []
  if (tree !== undefined) {
    result.push(tree.value)
    preOrderTraverse(tree.left, result)
    preOrderTraverse(tree.right, result)
  }
  return result
}

// 2.	中序优先遍历LDR递归版
function inOrderTraverse(tree, result) {
  result = result ? result : []
  if (tree !== undefined) {
    inOrderTraverse(tree.left, result)
    result.push(tree.value)
    inOrderTraverse(tree.right, result)
  }
  return result
}

// 3.	后序优先遍历LRD递归版
function postOrderTraverse(tree, result) {
  result = result ? result : []
  if (tree !== undefined) {
    postOrderTraverse(tree.left, result)
    postOrderTraverse(tree.right, result)
    result.push(tree.value)
  }
  return result
}

// 4. 广度优先(层级遍历)，自左往右，利用队列暂存数据
function levelOrder(tree) {
  const result = []
  const queue = []
  if (tree !== undefined) {
    queue.push(tree)
    while (queue.length) {
      tree = queue.shift()
      result.push(tree.value)
      // append children of current Node
      if (tree.left !== undefined) {
        queue.push(tree.left)
      }
      if (tree.right !== undefined) {
        queue.push(tree.right)
      }
    }
  }
  return result
}

// 5. 深度遍历(先序优先非递归版)
function preOrderUnRecursive(tree) {
  const result = []
  const queue = []
  if (tree !== undefined) {
    while (queue.length || tree) {
      if (tree) {
        result.push(tree.value)
        queue.push(tree)
        tree = tree.left
      } else {
        tree = queue.pop()
        tree = tree.right
      }
    }
  }
  return result
}

// 6. 深度遍历(中序优先非递归版)
function inOrderUnRecursive(tree) {
  var result = []
  var queue = []
  if (tree !== undefined) {
    while (queue.length || tree) {
      if (tree) {
        queue.push(tree)
        tree = tree.left
      } else {
        tree = queue.pop()
        result.push(tree.value)
        tree = tree.right
      }
    }
  }
  return result
}

// 7. 深度遍历(后序优先非递归版)
function postOrderUnRecursive(tree) {
  const result = []
  const queue = []
  var tmp
  if (tree !== undefined) {
    queue.push(tree)
    while (queue.length) {
      // 先得到最后一项作为比较项
      tmp = queue[queue.length - 1]
      // 把左子树按深度全部追加到queue
      if (tmp.left && tree !== tmp.left && tree !== tmp.right) {
        queue.push(tmp.left)
        // 把右子树追加到queue
      } else if (tmp.right && tree !== tmp.right) {
        queue.push(tmp.right)
      } else {
        // 左子树到最底层节点，开始打印left与right
        result.push(queue.pop().value)
        //  指向上一个节点，直到queue为空，也就是遍历至根节点
        tree = tmp
      }
    }
  }
  return result
}

/* === test === */

// 二叉树数据
const data = `
         1
      /      \
    2          3
  /   \      /    \
 4     5    6      7
      /  \
     8    9
`
console.log(data)
const tree = {
  value: 1,
  left: {
    value: 2,
    left: {
      value: 4
    },
    right: {
      value: 5,
      left: {
        value: 8
      },
      right: {
        value: 9
      }
    }
  },
  right: {
    value: 3,
    left: {
      value: 6
    },
    right: {
      value: 7
    }
  }
}
console.log(tree)

var result = []

// 1. 先序优先遍历 DLR
result = preOrderTraverse(tree)
//  [1, 2, 4, 5, 8, 9, 3, 6, 7]
console.log('1. preOrderTraverse:', result)

// 2. 中序优先遍历 LDR
result = inOrderTraverse(tree)
// [4, 2, 8, 5, 9, 1, 6, 3, 7]
console.log('2. inOrderTraverse:', result)

// 3. 后序优先遍历 LRD
result = postOrderTraverse(tree)
//  [4, 8, 9, 5, 2, 6, 7, 3, 1]
console.log('3. postOrderTraverse:', result)

// 4. 广度优先遍历
result = levelOrder(tree)
// [1, 2, 3, 4, 5, 6, 7, 8, 9]
console.log('4. levelOrder:', result)

// 5. 深度遍历（先序非递归）
result = preOrderUnRecursive(tree)
// [1, 2, 4, 5, 8, 9, 3, 6, 7]
console.log('5. preOrderUnRecursive:', result)

// 6. 深度遍历（中序非递归）
result = inOrderUnRecursive(tree)
// [4, 2, 8, 5, 9, 1, 6, 3, 7]
console.log('6. inOrderUnRecursive:', result)

// 7. 深度遍历（后序非递归）
result = postOrderUnRecursive(tree)
// [4, 8, 9, 5, 2, 6, 7, 3, 1]
console.log('7. postOrderUnRecursive:', result)

/**
jarry@jarrys-MacBook-Pro binarytree % node binary_tree.js

         1
      /         2         3
  /         /     4     5    6      7
      /       8    9

{ value: 1,
  left:
   { value: 2,
     left: { value: 4 },
     right: { value: 5, left: [Object], right: [Object] } },
  right: { value: 3, left: { value: 6 }, right: { value: 7 } } }
preOrderTraverse: [ 1, 2, 4, 5, 8, 9, 3, 6, 7 ]
inOrderTraverse: [ 4, 2, 8, 5, 9, 1, 6, 3, 7 ]
postOrderTraverse: [ 4, 8, 9, 5, 2, 6, 7, 3, 1 ]
levelOrder: [ 1, 2, 3, 4, 5, 6, 7, 8, 9 ]
preOrderUnRecursive: [ 1, 2, 4, 5, 8, 9, 3, 6, 7 ]
inOrderUnRecursive: [ 4, 2, 8, 5, 9, 1, 6, 3, 7 ]
postOrderUnRecursive: [ 4, 8, 9, 5, 2, 6, 7, 3, 1 ]
 */

File: 03-classic-algorithms/tree/binarytree/binary_tree.go
/**
 * Copyright © https://github.com/jarry All rights reserved.
 * @author: jarryli@gmail.com
 * @version: 1.0
 */

package main

import (
  "fmt"
)

// TreeNode 定义二叉树节点
type TreeNode struct {
  Value int
  Left  *TreeNode
  Right *TreeNode
}

// 1. 先序优先遍历 DLR 递归版
func preOrderTraverse(tree *TreeNode, result *[]int) {
  if tree != nil {
    *result = append(*result, tree.Value)
    preOrderTraverse(tree.Left, result)
    preOrderTraverse(tree.Right, result)
  }
}

// 2. 中序优先遍历 LDR 递归版
func inOrderTraverse(tree *TreeNode, result *[]int) {
  if tree != nil {
    inOrderTraverse(tree.Left, result)
    *result = append(*result, tree.Value)
    inOrderTraverse(tree.Right, result)
  }
}

// 3. 后序优先遍历 LRD 递归版
func postOrderTraverse(tree *TreeNode, result *[]int) {
  if tree != nil {
    postOrderTraverse(tree.Left, result)
    postOrderTraverse(tree.Right, result)
    *result = append(*result, tree.Value)
  }
}

// 4. 广度优先遍历（层级遍历）
func levelOrder(tree *TreeNode) []int {
  var result []int
  var queue []*TreeNode
  if tree != nil {
    queue = append(queue, tree)
    for len(queue) > 0 {
      node := queue[0]
      queue = queue[1:]
      // append children of current Node to result
      result = append(result, node.Value)
      if node.Left != nil {
        queue = append(queue, node.Left)
      }
      if node.Right != nil {
        queue = append(queue, node.Right)
      }
    }
  }
  return result
}

// 5. 深度遍历（先序非递归版）
func preOrderUnRecursive(tree *TreeNode) []int {
  var result []int
  var stack []*TreeNode
  for tree != nil || len(stack) > 0 {
    if tree != nil {
      result = append(result, tree.Value)
      stack = append(stack, tree)
      tree = tree.Left
    } else {
      tree = stack[len(stack)-1]
      stack = stack[:len(stack)-1]
      tree = tree.Right
    }
  }
  return result
}

// 6. 深度遍历（中序非递归版）
func inOrderUnRecursive(tree *TreeNode) []int {
  var result []int
  var stack []*TreeNode
  for tree != nil || len(stack) > 0 {
    if tree != nil {
      stack = append(stack, tree)
      tree = tree.Left
    } else {
      tree = stack[len(stack)-1]
      stack = stack[:len(stack)-1]
      result = append(result, tree.Value)
      tree = tree.Right
    }
  }
  return result
}

// 7. 深度遍历（后序非递归版）
// 后序遍历顺序：左子树 -> 右子树 -> 根节点。
// 使用栈模拟递归的过程，并通过一个变量 lastVisited 来记录上一次访问的节点。
func postOrderUnRecursive(tree *TreeNode) []int {
  var result []int          // 遍历结果
  var stack []*TreeNode     // 存储节点的栈
  var lastVisited *TreeNode // 最后访问的节点

  // 当前节点不为空或者栈不为空时，继续遍历，否则遍历完成
  for tree != nil || len(stack) > 0 {
    if tree != nil {
      // 如果当前节点不为空，则将其压入栈并遍历左子树
      stack = append(stack, tree)
      tree = tree.Left
    } else {
      // 查看栈顶节点（但不弹出）
      peekNode := stack[len(stack)-1]

      // 如果右子树存在，且右子树还未被访问过
      if peekNode.Right != nil && lastVisited != peekNode.Right {
        // 切换到右子树进行遍历
        tree = peekNode.Right
      } else {
        // 如果右子树不存在，或者右子树已经遍历过，则访问当前节点
        result = append(result, peekNode.Value)

        // 记录当前节点为最后访问的节点
        lastVisited = peekNode

        // 弹出当前节点
        stack = stack[:len(stack)-1]
      }
    }
  }

  // 返回后序遍历结果
  return result
}

/* === test === */

// 测试代码
func main() {
  // 二叉树数据
  /*
             1
         /      \
       2          3
     /   \      /    \
     4     5    6      7
         /  \
         8    9
  */
  tree := &TreeNode{
    Value: 1,
    Left: &TreeNode{
      Value: 2,
      Left:  &TreeNode{Value: 4},
      Right: &TreeNode{
        Value: 5,
        Left:  &TreeNode{Value: 8},
        Right: &TreeNode{Value: 9},
      },
    },
    Right: &TreeNode{
      Value: 3,
      Left:  &TreeNode{Value: 6},
      Right: &TreeNode{Value: 7},
    },
  }

  var result []int

  // 1. 先序优先遍历 DLR
  preOrderTraverse(tree, &result)
  //  [1, 2, 4, 5, 8, 9, 3, 6, 7]
  fmt.Println("1. preOrderTraverse:", result)

  // 2. 中序优先遍历 LDR
  result = []int{}
  inOrderTraverse(tree, &result)
  // [4, 2, 8, 5, 9, 1, 6, 3, 7]
  fmt.Println("2. inOrderTraverse:", result)

  // 3. 后序优先遍历 LRD
  result = []int{}
  postOrderTraverse(tree, &result)
  //  [4, 8, 9, 5, 2, 6, 7, 3, 1]
  fmt.Println("3. postOrderTraverse:", result)

  // 4. 广度优先遍历
  // [1, 2, 3, 4, 5, 6, 7, 8, 9]
  fmt.Println("4. levelOrder:", levelOrder(tree))

  // 5. 深度遍历（先序非递归）
  // [1, 2, 4, 5, 8, 9, 3, 6, 7]
  fmt.Println("5. preOrderUnRecursive:", preOrderUnRecursive(tree))

  // 6. 深度遍历（中序非递归）
  // [4, 2, 8, 5, 9, 1, 6, 3, 7]
  fmt.Println("6. inOrderUnRecursive:", inOrderUnRecursive(tree))

  // 7. 深度遍历（后序非递归）
  // [4, 8, 9, 5, 2, 6, 7, 3, 1]
  fmt.Println("7. postOrderUnRecursive:", postOrderUnRecursive(tree))
}

/**
jarry@MacBook-Pro binarytree % go run ./binary_tree.go
1. preOrderTraverse: [1 2 4 5 8 9 3 6 7]
2. inOrderTraverse: [4 2 8 5 9 1 6 3 7]
3. postOrderTraverse: [4 8 9 5 2 6 7 3 1]
4. levelOrder: [1 2 3 4 5 6 7 8 9]
5. preOrderUnRecursive: [1 2 4 5 8 9 3 6 7]
6. inOrderUnRecursive: [4 2 8 5 9 1 6 3 7]
7. postOrderUnRecursive: [4 8 9 5 2 6 7 3 1]
*/
